File: test/easy/check-perfect-number.test.ts
import { describe, test, expect } from "bun:test";
import {
    checkPerfectNumber,
    classifyPerfectNumber,
} from "@/src/easy/check-perfect-number";

describe("checkPerfectNumber", () => {
    test.each([
        ["496 is a perfect number", "perfect", 496, ""],
        ["8128 is a perfect number", "perfect", 8128, ""],
        ["8589869056 is a perfect number", "perfect", 8589869056, ""],
        ["137438691328 is a perfect number", "perfect", 137438691328, ""],

        [
            "Smallest perfect number is classified correctly",
            "perfect",
            6,
            "none",
        ],
        [
            "Medium perfect number is classified correctly",
            "perfect",
            28,
            "none",
        ],
        [
            "Large perfect number is classified correctly",
            "perfect",
            33550336,
            "none",
        ],

        [
            "Smallest abundant number is classified correctly",
            "abundant",
            12,
            "none",
        ],
        [
            "Medium abundant number is classified correctly",
            "abundant",
            30,
            "none",
        ],
        [
            "Large abundant number is classified correctly",
            "abundant",
            33550335,
            "none",
        ],

        [
            "Smallest prime deficient number is classified correctly",
            "deficient",
            2,
            "none",
        ],
        [
            "Smallest non-prime deficient number is classified correctly",
            "deficient",
            4,
            "none",
        ],
        [
            "Medium deficient number is classified correctly",
            "deficient",
            32,
            "none",
        ],
        [
            "Large deficient number is classified correctly",
            "deficient",
            33550337,
            "none",
        ],

        // ["Zero is rejected (not a natural number)", 0, "Classification is only possible for natural numbers."],
        // ["Negative integer is rejected (not a natural number)", -1, "Classification is only possible for natural numbers."],
    ])(
        "\n\tGiven: %s",
        (criteria: string, expected: string, num: number, err: string) => {
            const value = classifyPerfectNumber(num);
            const result = checkPerfectNumber(num);
            expect(value).toStrictEqual(expected);
            expect(result).toBe(value === "perfect");
        },
    );
});

File: test/easyswift/CheckPerfectNumberTest.swift
import Testing

@testable import easy

@Test
func checkPerfectNumberSkipTest() {
  #expect(true == true)
}
 
// @Test(arguments: [
//   (496, 496),
//   (8128, 8128),
//   // // (8_589_869_056, 8_589_869_056),
//   // // (137_438_691_328, 137_438_691_328),
//   (6, 6),
//   (28, 28),
//   // // (33_550_336, 33_550_336),
//   (12, 16),
//   (30, 42),
//   // (33_550_335, 39_885_825),
//   (2, 1),
//   (4, 3),
//   (32, 31),
//   (33_550_337, 1),
// ])
// func calculatePerfectNumberTest(n: UInt64, expected: UInt64) {
//   #expect(calculatePerfectNumber(n) == expected)
// }

// @Test(arguments: [
//   ("perfect", 496),
//   ("perfect", 8128),
//   // ("perfect", 8_589_869_056),
//   // ("perfect", 137_438_691_328),
//   ("perfect", 6),
//   ("perfect", 28),
//   // ("perfect", 33_550_336),
//   ("abundant", 12),
//   ("abundant", 30),
//   ("abundant", 33_550_335),
//   ("deficient", 2),
//   ("deficient", 4),
//   ("deficient", 32),
//   ("deficient", 33_550_337),
// ])
// func classifyPerfectNumberTest(expected: String, n: UInt64) {
//   #expect(classifyPerfectNumber(n) == expected)
// }
